### DISCUSSION TITLE
UDP_RECV_MAX_COALESCED_SIZE

### DISCUSSION THREAD

**Discussion by Emjayen** [microsoft/msquic#3494] [Q&A] [answered]
Regarding this option, of which I understand the requirements of this project were the impetus for its implementation, are there any particular requirements for it to be engaged? I haven't been able to coerce it to do so.

The test is straightforward:

- Windows 10/19044 (x64)
- Bound and connected UDP (IPv4) socket to explicit src/dst address.
- UDP_RECV_MAX_COALESCED_SIZE set to 64K (varied sizes, no change)
- AFD buffering disabled.
- Varying IRP queue depth (32-2048 tested), each with a single 64K buffer (varied sizes, no change)
- Wired back-to-back (10G) physical machines; sender tx rate of about 1.4Mpps (1024B frames)
- IOCP notification
- Tested pinning rx thread to same RSS queue directed CPU; no change.

Yet still haven't observed any coalescing; being woken up for 982B datagrams.

Thanks,

Matt.

**Comment by nibanks** [comment, marked as answer]
Please see https://github.com/microsoft/msquic/blob/main/docs/TSG.md#diagnosing-software-udp-receive-offload-issues. I believe URO might be getting disabled on your machine. There are known issues on Windows client where certain things on the system (i.e. Windows defender) can cause URO to get disabled. I recommend trying on Windows Server.

**Reply by Emjayen** [reply]
That was indeed all it was. Thank you.

A suggestion: you may wish to note in that documentation, in regards to performance issues: WFP callouts, if present, will cause a catastrophic degradation in performance; a single rx flow of negligible size (1Mpps; ~30k ISR/s) saturating an entire core. Removing WFP callouts from the path drops load by more than an order of magnitude.

Before:
![image](https://user-images.githubusercontent.com/12062565/224524248-2dfe4d16-7dcd-4b32-bba2-3051622e5ea4.png)

After:
![image](https://user-images.githubusercontent.com/12062565/224524257-1d1293c1-40da-480a-95b6-03f6089fe2c2.png)
